pkgs/applications/virtualization/singularity/generic.nix

```diff
@@ -199,8 +220,19 @@ in
     patchShebangs --build "$configureScript" makeit e2e scripts mlocal/scripts
 
     # Patching the hard-coded defaultPath by prefixing the packages in defaultPathInputs
-    substituteInPlace cmd/internal/cli/actions.go \
-      --replace "defaultPath = \"${defaultPathOriginal}\"" "defaultPath = \"''${defaultPathInputs// /\/bin:}''${defaultPathInputs:+/bin:}${defaultPathOriginal}\""
+    ${concatMapStringAttrsSep "\n" (fileName: originalDefaultPaths: ''
+      substituteInPlace ${lib.escapeShellArg fileName} \
+        ${
+          lib.concatMapStringsSep " \\\n  " (
+            originalDefaultPath:
+            lib.concatStringsSep " " [
+              "--replace-fail"
+              (addShellDoubleQuotes (lib.escapeShellArg originalDefaultPath))
+              (addShellDoubleQuotes ''$inputsDefaultPath''${inputsDefaultPath:+:}${lib.escapeShellArg originalDefaultPath}'')
+            ]
+          ) originalDefaultPaths
+        }
+    '') sourceFilesWithDefaultPaths}
```

Comment:
This is a bit involved and will eventually require refactoring... I think at this point we really should consider switching from templating to `__structuredAttrs`.

Reply:
This will assign our new "default paths" a higher priority than the FHS (`/usr/bin`) paths?

Reply:
> This is a bit involved and will eventually require refactoring...

I agree.

It would be much simpler if we only had one default path in one file to patch. This could be achieved by either
1. Improve the `defaultPath` specification by submitting a PR upstream. That sounds like a nontrivial fix, especially as I'm unfamiliar with Go.
2. Pick a file whose `defaultPath` most influences the choice of core privileged utilities. This requires further inspection of the impact of each `defaultPath` and would be prone to further fixes.



> This will assign our new "default paths" a higher priority than the FHS (`/usr/bin`) paths?

Yes, and this is the current behavior. It will change in #306730.